xwiki-platform-core/xwiki-platform-livedata/xwiki-platform-livedata-webjar/src/main/webjar/Logic.js

```diff
@@ -126,6 +164,277 @@ define(["jquery"], function ($) {
 
 
 
+
+  /**
+   * Return the property descriptor corresponding to a property id
+   * @param {String} propertyId
+   * @returns {Object}
+   */
+  Logic.prototype.getPropertyDescriptor = function (propertyId) {
+    var propertyDescriptor = null;
+    // Unfortunately the Array.prototype.find method is not supported in IE :(
+    this.data.meta.propertyDescriptors.some(function (descriptor) {
+    if (descriptor.id === propertyId) {
+        propertyDescriptor = descriptor;
+        return true;
+      }
+    });
+    return propertyDescriptor;
+  };
+
+
+
+  /**
+   * Return a new displayer based on the specified property and row data
+   * @param {String} propertyId The id of the property of the entry
+   * @param {Object} entry The entry data object
+   * @param {String} displayerId The
+   */
+  Logic.prototype.createDisplayer = function (propertyId, entry, displayerId) {
+    var self = this;
+    var defer = $.Deferred();
+
+    // default displayerId
+    if (displayerId === undefined) {
+      displayerId = (this.getPropertyDescriptor(propertyId).displayer || {}).id || "default";
+    }
+
+    // load displayer based on it's id
+    require([BASE_PATH + "displayers/" + displayerId + "Displayer.js"],
+      // load success
+      function (Displayer) {
+        var displayer = new Displayer(propertyId, entry, self);
+        defer.resolve(displayer);
+      },
+
+      // load failure
+      function (err) {
+        // try to load default layout instead
+        if (displayerId !== "default") {
+          self.createDisplayer(propertyId, entry, "default").then(function (displayer) {
+            defer.resolve(displayer);
+          }, function () {
+            defer.reject();
+          });
+        } else {
+          console.error(err);
+          defer.reject();
+        }
+      });
+
+      return defer.promise();
+  };
+
+
+
+
+  /**
+   * Update sort configuration based on parameters, then fetch new data
+   * @param {String} property The property to sort according to
+   * @param {String} level The sort level of the property (0 is the highest).
+   *   Undefined means current. Negative value removes property sort.
+   * @param {String} descending Specify whether the sort should be descending or not.
+   *   Undefined means toggle current direction
+   */
+  Logic.prototype.sort = function (property, level, descending) {
```

Comment:
Yes I thought about that, but for me it's the cases where you want to enforce the direction are also pretty rare, most of the time you want to toggle it or get the default one for a new column.
Both makes sense for me, but I as think that the sort level has a "higher importance" (you first decide which which property to sort, then to what level, then to what direction), I decided to put it before.

Reply:
> Then, with the current API if I want to sort on 2 columns I need to make 2 sort() calls which will generate two HTTP requests to fetch the data.

When the user will want to sort two columns with different levels, that's basically how he is going to do, specifying one sort after the other. As fetching new data is automatic when the user modify the Livedata configuration, there will be two HTTP calls anyway. So I think the API should reflect the way the user behave, instead of relying on a component to pass the information in the correct way.
Of course, if we want to specify several sorts at once, for instance during initialization, we can create another function in the API that take the whole sort state, and update the configuration accordingly (that's what I proposed in another review of this PR).

Reply:
> When the user will want to sort two columns with different levels, that's basically how he is going to do, specifying one sort after the other.

This doesn't imply that the live data needs to make an HTTP request after each change. The card layout could have a Sort **modal** where the user selects the sort columns, the direction, and the priority between the sort columns and then clicks the Sort button to close the Modal and apply the sort. The live data would make the HTTP request only at the end.

>  As fetching new data is automatic when the user modify the Livedata configuration, there will be two HTTP calls anyway.

I don't see why you can't modify all sort columns at once.